src/analyses/weaklyRelationalPointerAnalysis.ml

```diff
@@ -0,0 +1,186 @@
+(** A Weakly-Relational Pointer Analysis. The analysis can infer equalities and disequalities between terms which are built from pointer variables, with the addition of constants and dereferencing. ([wrpointer])*)
+
+open Analyses
+open GoblintCil
+open WeaklyRelationalPointerDomain
+module CC = CongruenceClosure
+open CC.CongruenceClosure
+open Batteries
+
+module Spec =
+struct
+  include DefaultSpec
+  include Analyses.IdentitySpec
+  module D = D
+  module C = D
+
+  let name () = "wrpointer"
+  let startcontext () = D.empty ()
+
+  (* find reachable variables in a function *)
+  let reachable_from_args ctx args =
+    let res =
+      List.fold (fun vs e -> vs @ (ctx.ask (ReachableFrom e) |> Queries.AD.to_var_may)) [] args in
+    if M.tracing then M.tracel "wrpointer-reachable" "reachable vars: %s\n" (List.fold_left (fun s v -> s ^v.vname ^"; ") "" res); res
+
+  (* Returns Some true if we know for sure that it is true,
+     and Some false if we know for sure that it is false,
+     and None if we don't know anyhing. *)
+  let eval_guard ask t e =
+    let prop_list = T.prop_of_cil ask e true in
+    let res = match split prop_list with
+      | [], [] -> None
+      | x::xs, _ -> if fst (eq_query t x) then Some true else if neq_query t x then Some false else None
+      | _, y::ys ->  if neq_query t y then Some true else if fst (eq_query t y) then Some false else None
+    in if M.tracing then M.trace "wrpointer" "EVAL_GUARD:\n Actual guard: %a; prop_list: %s; res = %s\n"
+        d_exp e (show_conj prop_list) (Option.map_default string_of_bool "None" res); res
+
+  let query_may_point_to (ask:Queries.ask) t e =
+    match T.of_cil ask e with
+    | Some term, Some offset ->
+      begin match insert t term with
+        | _,None -> MayBeEqual.AD.top()
+        | _,Some cc ->
+          let comp = Disequalities.comp_t cc.uf term in
+          let valid_term (t,z) =
+            T.is_ptr_type (T.type_of_term t) && (T.get_var t).vid > 0 in
+          let equal_terms = List.filter valid_term comp in
+          let intersect_query_result res (term,z) =
+            let next_query = ask.f (MayPointTo (T.to_cil_sum z (T.to_cil term))) in
+            MayBeEqual.AD.meet res next_query in
+          List.fold intersect_query_result (MayBeEqual.AD.top()) equal_terms
+      end
+    | _ ->
+      MayBeEqual.AD.top()
```

Comment:
Could we as alternative indicator of where the domain is helpful somehow measure how often this leads to an improvement?
We would have to somehow measure this at the top level though...